src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/UpsertResourceHandler.cs

```diff
@@ -53,7 +53,12 @@ public async Task<UpsertResourceResponse> Handle(UpsertResourceRequest request,
         {
             EnsureArg.IsNotNull(request, nameof(request));
 
-            if (await AuthorizationService.CheckAccess(DataActions.Write, cancellationToken) != DataActions.Write)
+            // Check for create and update permissions, maintaining legacy Write support
+            var requiredAccess = DataActions.Create | DataActions.Update | DataActions.Write;
+            var grantedAccess = await AuthorizationService.CheckAccess(requiredAccess, cancellationToken);
+
+            // Need either specific granular permission or legacy Write permission
+            if ((grantedAccess & (DataActions.Create | DataActions.Update | DataActions.Write)) == 0)
```

Comment:
If a user has a permission to only update the resource, with this logic user can potentially Create the resource as well?

Reply:
Updated for create vs update